Source/WebCore/loader/FrameLoader.cpp

```diff
@@ -2791,10 +2791,15 @@ void FrameLoader::checkLoadCompleteForThisFrame()
             RefPtr document = m_frame->document();
             OptionSet<DataDetectorType> types = m_frame->settings().dataDetectorTypes();
             if (document && types) {
-                auto documentLevelResults = retainPtr(DataDetection::detectContentInRange(makeRangeSelectingNodeContents(*document), types, m_client->dataDetectionReferenceDate()));
-                m_frame->dataDetectionResults().setDocumentLevelResults(documentLevelResults.get());
-                if (m_frame->isMainFrame())
-                    m_client->dispatchDidFinishDataDetection(documentLevelResults.get());
+                DataDetection::detectContentInFrame(m_frame, types, m_client->dataDetectionReferenceDate(), [this, weakFrame = WeakPtr { m_frame.get() }, document](NSArray *results) {
+                    if (!weakFrame || &weakFrame->loader() != this)
```

Comment:
*FIRST* thing you should do before using weakFrame is store it is a RefPtr and then only use that RefPtr from now on.